From 33e2789588a87dedbc19ae7a0304c6012045f28b Mon Sep 17 00:00:00 2001
From: Kendall Willis <k-willis@ti.com>
Date: Wed, 12 Mar 2025 20:39:04 -0500
Subject: [PATCH 151/181] PENDING: pmdomain: ti_sci: Handle wakeup constraint
 if device has pinctrl wakeup state

Devices that are marked as a wakeup source and use pinctrl wakeup state
should not set wakeup constraints since these can happen even from deepest
low power state, so should not prevent the DM from picking deep power
states.

For example, certain TI SoCs support I/O Only + DDR low power mode in
which the entire SoC is off except I/O pins. The wakeup sources for this
mode include mcu_uart0, mcu_mcan0, mcu_mcan1, and wkup_uart0. If these
sources send a wakeup constraint to the DM, I/O Only + DDR would not be
picked by the DM as the low power state to go in. This patch provides a
path for the named wakeup sources to avoid setting a wakeup constraint
in order to enter I/O Only + DDR LPM.

Detect the pinctrl wakeup state in the suspend path, and if it exists,
skip sending the constraint.

Signed-off-by: Kendall Willis <k-willis@ti.com>
---
 drivers/pmdomain/ti/ti_sci_pm_domains.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/pmdomain/ti/ti_sci_pm_domains.c b/drivers/pmdomain/ti/ti_sci_pm_domains.c
index 8e320e1f3f49..c3cd25615516 100644
--- a/drivers/pmdomain/ti/ti_sci_pm_domains.c
+++ b/drivers/pmdomain/ti/ti_sci_pm_domains.c
@@ -16,6 +16,7 @@
 #include <linux/pm_runtime.h>
 #include <linux/slab.h>
 #include <linux/soc/ti/ti_sci_protocol.h>
+#include <linux/pinctrl/consumer.h>
 #include <dt-bindings/soc/ti,sci_pm_domain.h>
 
 /**
@@ -84,9 +85,24 @@ static inline void ti_sci_pd_set_wkup_constraint(struct device *dev)
 	struct generic_pm_domain *genpd = pd_to_genpd(dev->pm_domain);
 	struct ti_sci_pm_domain *pd = genpd_to_ti_sci_pd(genpd);
 	const struct ti_sci_handle *ti_sci = pd->parent->ti_sci;
+	struct pinctrl *pinctrl = devm_pinctrl_get(dev);
+	struct pinctrl_state *pinctrl_state_wakeup;
 	int ret;
 
 	if (device_may_wakeup(dev)) {
+		/*
+		 * If device can wakeup using pinctrl wakeup state,
+		 * we do not want to set a constraint
+		 */
+		if (!IS_ERR_OR_NULL(pinctrl)) {
+			pinctrl_state_wakeup = pinctrl_lookup_state(pinctrl, "wakeup");
+			if (!IS_ERR_OR_NULL(pinctrl_state_wakeup)) {
+				dev_dbg(dev, "%s: has wake pinctrl wakeup state, not setting " \
+						"constraints\n", __func__);
+				return;
+			}
+		}
+
 		/*
 		 * If device can wakeup using IO daisy chain wakeups,
 		 * we do not want to set a constraint.
-- 
2.49.0

